### DISCUSSION TITLE
✔ Tasks from June 28th 2022

### DISCUSSION THREAD

**Discussion by KennedyRichard** [IndiePython/nodezator#14] [Announcements]
Hello, everyone! Every time I'm about to dive into working on a new branch or couple of branches I'll create a discussion like this one with the following purposes:

- to announce the features I intend to implement in the new branch;
- to share with you my thoughts and insights on the changes to be implemented;
- to give you the opportunity to share your thoughts on this, suggest other changes, before I start working on the branch;
- once I start implementing these changes, to share my progress with you and share new insights into the problems;

# Changes planned for the next branches to start on 2022-June-28:
- [x] make title of data nodes editable;
- [x] make window resizable;

## thoughts on "make title of data nodes editable"

This should be a pretty quick implementation, since Nodezator already has all the utilities needed to implement this. I estimate it will take me 2 days to implement it; data nodes are used as variables within a node layout and it is very important and only natural that users can change their names.

## thoughts on "make window resizable"

This change should take a bit more time, because it is more complex. I estimate it will take me a week or two, because it will require me to change the in-app editors to be able to correctly reposition according to the dimensions used and also alter some caching mechanisms for the window contents (which might result in more time needed to implement the future). I'll also may need to implement additional measures to ensure performance is kept at acceptable levels because the default pygame graphics back end (like most libraries) has to work more the bigger the window.

With this in mind, here's the behaviour I want to implement: allowing the user to click the "maximize window" so the window occupies the entire space on the screen. This is the ideal solution. If everything goes smoothly and we keep performance acceptable, we'll stick with this solution. Otherwise, I'll present a new approach and pursue it.

## Your thoughts

Remember that I'll start to implement those changes on next June 28th. If you are ok with this, you don't need to do anything else (maybe a thumbs up, though?), but if you have anything to say, please share your thoughts so that we are on the same page.

Several users requested this feature, but Mr. Feng also started a issue requesting it (which I converted into a discussion since a feature request is not an issue). You can also take a look and comment on it if you want: #15 

## Other changes performed in addition to the planned ones:

- [x] add more context menus (https://github.com/IndiePython/nodezator/discussions/14#discussioncomment-3287448);
- [x] being able to replace operation in operator nodes;
- [x] move logs out of the app folder;

**Comment by KennedyRichard** [comment]
As planned, I'm starting working on the proposed changes today. I'll start with a new branch to make Nodezator window resizable. Estimated time of implementation is one or two weeks for reasons already listed above. But who knows, it may be ready sooner than expected. Expect updates here every other day as the implementation progresses.

Also remember that  my work on the Indie Python project requires that I spend time not only on development, but also on maintaining the websites, posting about Nodezator on social media and other online communities, occasionally writing and producing videos, etc. I'll usually prioritize development work, though, because it is the project's activity that generates more values for the general audience and patrons. After all, quality tools and a great community are the pillars of any open source project.

I wish you a great week!

**Comment by pcisar** [comment]
While I really appreciate the ability to resize the main window (btw, I think that main reason for this feature request was need to go full-screen, not ability to resize to any size), it does not solve the problem with limited workspace / canvas for building node graphs. I see as very serious issue / limitation that it's not possible to create graphs that does not fit the window. This significantly limits the complexity of graphs one can make, especially when nodes have multiple input/output sockets (with widgets) that occupy serious amount of space on screen.

This is certainly related to "window resize" feature, because what would nodezator do if one would create graph in full-screen and then resize to smaller window ? Right now, nodezator clips content that does not fit the screen.

EDIT: I'm sorry, I've just discovered that WASD scrolling (available to scroll off-screen content in text views like node source) also works for main window. Which sort-off answers my concerns. However, it's not intuitive (and I didn't remember mentioned this in documentation either) as there are no cues that there is any content off-screen and in which direction is. 

For example next sequence:
1. create a node and scroll it far off-screen with WASD
2. create a graph with other nodes
3. save the graph with main cluster in view (the first one is out of screen)
4. quit nodezator
5. start nodezator and load the file
6. You'll see main graph cluster with no cue that there is a node far away somewhere

**Reply by KennedyRichard** [reply]
Mr. Císař, thank your for sharing your thoughts/concerns.
 
> I'm sorry, I've just discovered that WASD scrolling (available to scroll off-screen content in text views like node source) also works for main window. Which sort-off answers my concerns.

No need to apologize for anything. The user inability to find an existing feature is solely my own fault as the creator and designer of the app/feature. I'm thankful for your initiative to share your problems/concerns.

Also, if you want to know more controls, they can all be found on the menu **Help > Controls**. Pressing <F1> will also bring a screen to allow you to see general controls and text editor controls.

> However, it's not intuitive (and I didn't remember mentioned this in documentation either)

You are right. I should've also included at least the arrow keys for scrolling as well. I'll do so immediately, as I believe this will help a bit and is super quick (just a few extra lines) to implement.

There's also a lot of features not discussed on the manual yet for the sole reason I didn't have time yet to include them. Another of my own shortcomings. However, rest assured I'm committed to address every and each issue as it appears. It is just that I schedule my priority tasks according, among other things, to user feedback. Now that you reached me, I'll take your concerns into account when scheduling my next actions.

> there are no cues that there is any content off-screen and in which direction is

You are also right on this one. For now, the only feature that can help on this to some extent, is the graph snap feature: <Shift+Numpad digit from 1 to 9>, which snaps (moves) the screen to one of the corners of the node layout.

For instance, <Shift+Numpad 5> moves the screen to the center of the node layout (since 5 is in the center of the numpad), <Shift + Numpad 6> goes to the midright position of the node layout (since 6 is on the midright corner of the numpad) and so on. This helps a bit because, for instance, if you go to the topleft corner of the node layout (<Shift + Numpad 7>), you can be sure there are no objects above or to the left of that position, since you'll already be on the topleft corner of the layout.

However, although this helps, I don't think it is enough. Here's another thing I'm thinking of implementing: an overview/map view feature (probably using the <m> key), that lets you see all a simplified and minimized version of all objects on the node layout at once as a map on the screen, and let's you move the screen around as well.

Here's a sketch of  the map view feature below. The whole image represents all objects in your node layout and the rectangle outline represents what is visible on screen, and you can move the screen to any part of the node layout by clicking on the image.

![map view feature sketch](https://i.imgur.com/IS8f6wN.png)

I'm also thinking of implementing a list view, which would list all existing objects by id and name, and let you snap (move) the screen to any of them by clicking the item on the list.

> For example next sequence

Also thank you for providing so much detail. It always helps, and I could clearly understand your frustration by following the steps provided on my mind.

---

I think this address your main concerns, but please, if I missed anything don't hesitate to reach me again. Also please, don't hesitate to share any additional suggestions you may have as well.

**Reply by KennedyRichard** [reply]
Hello, Mr. @pcisar ! It's been quite some time since we last spoke! Nodezator's development has been active since its release, with no plans to slow down or stop. Today I'm here just to say I finally found the time to implement this map/overview feature I mentioned to you in my reply above. I thought you'd like to know, regardless of whether you still employ Nodezator in your workflow or not.

It's available in the new 1.5 version released yesterday. The feature is called bird's eye view and allows the user to both see the entire graph on the screen and to jump to any spot on it instantaneously. Here's a demonstration of its usage from the devlog (https://github.com/IndiePython/nodezator/discussions/72#discussioncomment-9470576):

https://private-user-images.githubusercontent.com/18728298/331599052-268d110e-c6e1-4e8a-b273-3eca864f8882.mp4

Nodezator still has a long, long way to go, but little by little we are getting there.

Have a nice weekend!

**Comment by KennedyRichard** [comment]
Hello, everyone, as I said, I would be providing some update here on the progress of these next branches every other day.

The window resize feature implementation is going very smoothly. The only reason this implementation takes more time is the sheer amount of small adjustments I have to make in a lot of different modules. But apart from that the changes are relatively simple, though they usually require a lot of testing and fine-tuning.

I'd also like the opportunity to make clear that though this feature will be super handy for everyone as it will allow everyone to increase the amount of screen used for node editing, those planning to work on super big screens (4k and higher) will probably not be able to use the full dimensions of their screen, as the app would not perform well on such dimensions.

However, you'll still benefit, as you can still increase the amount of screen used. This is not a limitation exclusive to pygame, as 4k screens and higher are usually very resource intensive for any app. Also, I expect this limitation to be surpassed when we switch the Nodezator graphics backend to OpenGL some time in the future.

That's all for now. Peace.

**Comment by KennedyRichard** [comment]
Reporting my work

![](https://media.giphy.com/media/LmZmWJk6GtN9FSZpOv/giphy.gif)

Resizable window implementation still going smoothly. In case some of you didn't know, you can observe the changes by comparing the master and resizable-window branches https://github.com/KennedyRichard/nodezator/compare/master...resizable-window. As I told you before, most changes are relatively small, but there are many (63 files changed so far), and they require testing/fine-tuning. I should be finishing and merging the changes in a few days.

**Comment by KennedyRichard** [comment]
Hello, everyone! Almost done with the resizable window feature! I actually finished it, I just want a bit of time for testing. I should merge with master and upload to pypi by tomorrow! :hourglass_flowing_sand:

**Comment by KennedyRichard** [comment]
Finally finished testing and debugging the resizable window feature! Just merged the resizable-window branch into master, pushed and uploaded the new Nodezator version to pypi! To install/upgrade to the new version, just execute

```bash
pip3 install --upgrade nodezator
```
This should make it easier to work with Nodezator in big screens, but remember that the larger the screen, the higher the impact on performance. If you plan to use Nodezator in 4K screens, be aware that you might not be able to use it efficiently at that size. Even so, you'll at least **be able to make Nodezator bigger than the previously fixed 1280x720 size**.

There are still stuff that can be improved, but it is all aesthetics, I'll come back to it eventually. Functionally everything should be fine. If you find any bugs, please report.

Thanks for the patience, it took a few days more than I thought it would (almost 03 weeks instead of the expected 02 weeks), but it was all due to my other tasks on the Indie Python project and solving last minute issues (like the one closed 02 days ago).

The next change planned, as listed in the top of this discussion, is to **make the title of data nodes editable**. Data nodes represent variables you define within your node layout, so it is only natural that you should be able to name then as you see fit.

**Comment by KennedyRichard** [comment]
Though the resizable window feature is implemented, I realized I have to update the viewer nodes in the example node packs (the ones using pillow and matplotlib). This will still take me a couple of days, which is why I didn't implement the edition of data nodes' titles yet. Peace

Edit: must also update the example code in the manual section teaching how to make viewer nodes (https://manual.nodezator.com/ch02-defining-viewer-nodes.html)

**Comment by KennedyRichard** [comment]
Finished updating the viewer nodes on the matplotlib and pillow examples yesterday. Now, they behave properly when resizing the window and I even added support for scrolling the visuals with the keyboard or by click-dragging with the mouse.

Now I just need to finish updating the manual so it shows how to integrate window resizing when creating viewer nodes.

Once this is done, I'll finally be able to make the title on data nodes editable and create a new discussion to plan which items from the list of planned/requested features should be implemented next.

Peace

**Comment by KennedyRichard** [comment]
> Now I just need to finish updating the manual so it shows how to integrate window resizing when creating viewer nodes.
>
> Once this is done, I'll finally be able to make the title on data nodes editable...

Finished updating the manual and already uploaded the new version to pypi. Updated the in-app version as well as the web version of the manual.

Now the manual has a new chapter demonstrating how to define a more complex viewer node whose visualization adjusts when the window is resized and which allows the user to move it using both the keyboard and click-dragging with the mouse when the visualization gets bigger than the screen.

Here's the link for the new chapter on the web version: https://manual.nodezator.com/ch-defining-complex-viewer-nodes.html

Now, as I said before, I'll start working on making the title of data nodes editable, which should take a few days.

**Comment by KennedyRichard** [comment]
Update: since editing the data nodes' titles will require a context menu (a popup menu, by right-clicking), I decided to use the opportunity to implement context menus for the other objects in the graph as well, that is, the other kind of nodes and the text block as well. This will take a few more days anticipated, but is proceeding smoothly.

Here's a preview of the context menus when right-clicking a text block in two different circumstances: when it is not part of a selection (left) and when it is part of a selection (right).

![context menu screenshots](https://i.imgur.com/906oe8G.png)

**Comment by KennedyRichard** [comment]
Hello, everyone, just released Nodezator 1.2 yesterday.

![new information gif](https://media3.giphy.com/media/U23WekMlGy6cImpMim/giphy.gif?cid=790b76112da495a797440a82374ec4f0fe927fca295c79a5&rid=giphy.gif&ct=g)

The version is a lot friendlier than the past ones since objects in the graph now have context menus.

You can finally edit the title of data nodes as promised.

Also, operations on operator nodes can now be replaced very easily by simply picking the new operation on its context menu. That is, you don't need to delete the node, create a new one and reconnect the sockets anymore. Everything is performed automatically for you, as soon as you pick the new operation you want.

All kinds of nodes now have context menus and the text blocks have as well.

I just need to implement context menus for the sockets and maybe for the widgets, then we'll finally be able to schedule and start working the next features to be implemented.

I attached a short video containing a quick demonstration of the new features.

Thank you all for your patience.

https://user-images.githubusercontent.com/18728298/183928900-e9d51127-f1fc-43ec-9a9b-1ff8f6bf55b6.mp4

**Comment by KennedyRichard** [comment]
Hello, everyone!

All the listed changes were finally implemented.

![lord of the rings frodo saying "it's done"](https://media0.giphy.com/media/3oKIPf3C7HqqYBVcCk/giphy.gif?cid=790b7611ebac0d8ae0212835fc66015cc6be185d0423965a&rid=giphy.gif&ct=g)

As I said before, I'll now check the pending issue and pull requests. After this I'll create a new discussion just like this one, where I'll propose, explain and gather feedback regarding the next planned/requested changes to be implemented.

Thanks for your patience.

**Comment by KennedyRichard** [comment]
Forgot to close the discussion. Doing it now!

Peace
